File: google_investointi_2027_2028/malli.R
# Yksinkertainen panos-tuotos (Leontief) -laskentamoottori.
#
# Kayttotarkoitus: kun toimialoittainen tekninen kerroinmatriisi A (vain
# kotimaiset, tuonnista puhdistetut valituotepanokset) ja toimialoittaiset
# arvonlisayskertoimet on saatu Tilastokeskuksen tarjonta- ja
# kayttotaulukoista (ks. hae_tilastokeskus.R), tama moduuli ratkaisee
# kokonaistuotannon ja siita lasketun BKT (arvonlisays) -vaikutuksen
# annetulle kysyntashokkivektorille.
#
# Kayttaa base R:n solve()-funktiota lineaariyhtaloryhman ratkaisuun -
# ei tarvitse ulkoisia paketteja.

#' Kokonaistuotanto x = (I - A)^-1 %*% d
#'
#' @param A Toimialoittainen tekninen kerroinmatriisi (n x n)
#' @param kysynta Kysyntashokkivektori (pituus n)
#' @return Kokonaistuotantovektori (pituus n)
leontief_tuotanto <- function(A, kysynta) {
  n <- length(kysynta)
  if (!all(dim(A) == c(n, n))) {
    stop("A:n tulee olla n x n -matriisi, jossa n = kysynta-vektorin pituus.")
  }
  I <- diag(n)
  as.numeric(solve(I - A, kysynta))
}

#' BKT (arvonlisays) -vaikutus = arvonlisayskertoimet . x
#'
#' @param A Toimialoittainen tekninen kerroinmatriisi (n x n)
#' @param arvonlisays_kertoimet Arvonlisayskertoimet toimialoittain (pituus n)
#' @param kysynta Kysyntashokkivektori (pituus n)
#' @return list(bkt = kokonais-BKT-vaikutus, tuotanto = toimialoittainen
#'   kokonaistuotanto)
arvonlisays_vaikutus <- function(A, arvonlisays_kertoimet, kysynta) {
  x <- leontief_tuotanto(A, kysynta)
  bkt <- sum(arvonlisays_kertoimet * x)
  list(bkt = bkt, tuotanto = x)
}

File: google_investointi_2027_2028/malli.py
"""Yksinkertainen panos-tuotos (Leontief) -laskentamoottori.

Kayttotarkoitus: kun toimialoittainen tekninen kerroinmatriisi A (vain
kotimaiset, tuonnista puhdistetut valituotepanokset) ja toimialoittaiset
arvonlisayskertoimet on saatu Tilastokeskuksen tarjonta- ja kayttotaulukoista
(ks. hae_tilastokeskus.py), tama moduuli ratkaisee kokonaistuotannon ja siita
lasketun BKT (arvonlisays) -vaikutuksen annetulle kysyntashokkivektorille.

Ei riipu numpysta / pandasista - toimii pelkalla Python-standardikirjastolla,
koska matriisi on kooltaan pieni (muutamia-kymmenia toimialoja).
"""
from __future__ import annotations


def solve_linear_system(matrix: list[list[float]], vector: list[float]) -> list[float]:
    """Ratkaisee Mx = v Gaussin eliminaatiolla osittaisella pivotoinnilla."""
    n = len(vector)
    aug = [row[:] + [vector[i]] for i, row in enumerate(matrix)]

    for col in range(n):
        pivot_row = max(range(col, n), key=lambda r: abs(aug[r][col]))
        if abs(aug[pivot_row][col]) < 1e-12:
            raise ValueError("Matriisi on singulaarinen - malli ei ratkea.")
        aug[col], aug[pivot_row] = aug[pivot_row], aug[col]

        pivot = aug[col][col]
        aug[col] = [value / pivot for value in aug[col]]

        for r in range(n):
            if r != col:
                factor = aug[r][col]
                aug[r] = [aug[r][k] - factor * aug[col][k] for k in range(n + 1)]

    return [aug[i][n] for i in range(n)]


def leontief_output(A: list[list[float]], final_demand: list[float]) -> list[float]:
    """Kokonaistuotanto x = (I - A)^-1 * d."""
    n = len(final_demand)
    identity_minus_A = [
        [(1.0 if i == j else 0.0) - A[i][j] for j in range(n)] for i in range(n)
    ]
    return solve_linear_system(identity_minus_A, final_demand)


def value_added_impact(
    A: list[list[float]],
    value_added_coeff: list[float],
    final_demand: list[float],
) -> tuple[float, list[float]]:
    """Laskee BKT (arvonlisays) -vaikutuksen = arvonlisayskertoimet . x.

    Palauttaa (bkt_vaikutus_yhteensa, toimialoittainen_kokonaistuotanto).
    """
    x = leontief_output(A, final_demand)
    bkt = sum(value_added_coeff[i] * x[i] for i in range(len(x)))
    return bkt, x
